### PR TITLE
[SPARK-40251][BUILD][MLLIB] Upgrade dev.ludovic.netlib from 2.2.1 to 3.0.2 & breeze from 2.0 to 2.1.0

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr aim to 
> A.Upgrade dev.ludovic.netlib from 2.2.1 to 3.0.2
> B.Upgrade breeze from 2.0 to 2.1.0

FYI
image

### Why are the changes needed?
A.dev.ludovic.netlib
> This version contain bug fix: 
> https://github.com/luhenry/netlib/issues/12
> https://github.com/luhenry/netlib/issues/10

> The changes between 2.2.1 and 3.0.2 as follows:
> https://github.com/luhenry/netlib/compare/v2.2.1...v3.0.2

B.breeze
> The changes between 2.0 and 2.1.0 as follows:
> https://github.com/scalanlp/breeze/compare/releases/v2.0...releases/v2.1.0

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Pass GA & exist UT.

### CODE REVIEW

**Comment by LuciferYang** [comment]
cc @zhengruifeng FYI

**Comment by panbingkun** [comment]
i will run BLASBenchmark

File: mllib/src/main/scala/org/apache/spark/mllib/linalg/LAPACK.scala [hunk 1]
```diff
@@ -17,9 +17,9 @@
 
 package org.apache.spark.mllib.linalg
 
-import dev.ludovic.netlib.{JavaLAPACK => NetlibJavaLAPACK,
-                           LAPACK => NetlibLAPACK,
-                           NativeLAPACK => NetlibNativeLAPACK}
+import dev.ludovic.netlib.lapack.{JavaLAPACK => NetlibJavaLAPACK,
```
Comment (LuciferYang) [inline mllib/src/main/scala/org/apache/spark/mllib/linalg/LAPACK.scala:20]: The format of `imports` looks strange, I'm not sure if it's standard

Comment (panbingkun) [reply mllib/src/main/scala/org/apache/spark/mllib/linalg/LAPACK.scala]: I will change it to follow the general rules

File: mllib-local/src/test/java/dev/ludovic/netlib/blas/F2jBLASUtil.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ * This is a workaround since class {@link F2jBLAS} are not public
+ */
+public class F2jBLASUtil {
+
+    private F2jBLASUtil() {}
+
+    public static JavaBLAS getInstance() {
+        return F2jBLAS.getInstance();
+    }
+}
```
Comment (panbingkun) [inline mllib-local/src/test/java/dev/ludovic/netlib/blas/F2jBLASUtil.java:30]: https://github.com/luhenry/netlib/blob/master/blas/src/main/java/dev/ludovic/netlib/blas/F2jBLAS.java#L28-L36
image

File: pom.xml [hunk 3]
```diff
@@ -1080,7 +1079,7 @@
       <dependency>
         <groupId>org.scalanlp</groupId>
         <artifactId>breeze_${scala.binary.version}</artifactId>
-        <version>2.0</version>
```
Comment (zhengruifeng) [inline pom.xml:1083]: please also add the upgrade of breeze in PR title and description

**Review by zhengruifeng — APPROVED** [review]

**Comment by zhengruifeng** [comment]
also cc @srowen @luhenry

**Comment by LuciferYang** [comment]
I think we should update the benchmark result of `BLASBenchmark`

File: pom.xml (hunk 3 above)
Comment (panbingkun) [reply pom.xml:1083]: okay

**Comment by panbingkun** [comment]
> I think we should update the benchmark result of `BLASBenchmark`

Yes, i will do it!

File: mllib-local/src/test/java/dev/ludovic/netlib/blas/F2jBLASUtil.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package dev.ludovic.netlib.blas;
+
+/**
+ * This is a workaround since class {@link F2jBLAS} are not public
+ */
+public class F2jBLASUtil {
```
Comment (srowen) [inline mllib-local/src/test/java/dev/ludovic/netlib/blas/F2jBLASUtil.java:23]: I don't think this matters at all, but I'd say 'final' here

Comment (LuciferYang) [reply mllib-local/src/test/java/dev/ludovic/netlib/blas/F2jBLASUtil.java:23]: Can we use reflection to get `JavaBLAS` and move this class into a spark namespace package? Is this better than adding a `ludovic/netlib` package in Spark code?

Comment (srowen) [reply mllib-local/src/test/java/dev/ludovic/netlib/blas/F2jBLASUtil.java:23]: For just test purposes, the new package and shim seems fine, but I would be fine with either

**Comment by panbingkun** [comment]
Don't merge it! There is a problem with running BLASBenchmark, wired!
I will fix it later.

File: mllib-local/pom.xml [hunk 5]
```diff
@@ -61,6 +61,11 @@
       <groupId>org.apache.spark</groupId>
       <artifactId>spark-tags_${scala.binary.version}</artifactId>
     </dependency>
+    <dependency>
```
Comment (srowen) [inline mllib-local/pom.xml:64]: Oh, I don't think we want to introduce this circular dependency. What's it for?

**Review by srowen — CHANGES_REQUESTED** [review]

File: mllib-local/pom.xml (hunk 5 above)
Comment (panbingkun) [reply mllib-local/pom.xml:64]: for Utils.classForName

Comment (srowen) [reply mllib-local/pom.xml:64]: We can't do it for that. Just inline what you need. There may be a 'copy' in an MLlib utils class even, not sure

Comment (LuciferYang) [reply mllib-local/pom.xml:64]: > We can't do it for that. Just inline what you need. There may be a 'copy' in an MLlib utils class even, not sure

+1, Agree

**Review by srowen — APPROVED** [review]

**Review by luhenry — APPROVED** [review]
Thank you for doing that!

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**